rllib/algorithms/appo/tf/appo_tf_policy_rlm.py

```diff
@@ -0,0 +1,242 @@
+            2.0,
+        )
+        logp_ratio = is_ratio * tf.math.exp(
+            target_actions_logp_time_major - behaviour_actions_logp_time_major
+        )
+
+        clip_param = self.config["clip_param"]
+        surrogate_loss = tf.math.minimum(
+            pg_advantages * logp_ratio,
+            (
+                pg_advantages
+                * tf.clip_by_value(logp_ratio, 1 - clip_param, 1 + clip_param)
+            ),
+        )
+        action_kl = old_target_policy_dist.kl(target_policy_dist)
+        mean_kl_loss = tf.math.reduce_mean(action_kl)
+        mean_pi_loss = -tf.math.reduce_mean(surrogate_loss)
+
+        # The baseline loss.
+        delta = values_time_major - vtrace_adjusted_target_values
+        mean_vf_loss = 0.5 * tf.math.reduce_mean(delta**2)
+
+        # The entropy loss.
+        mean_entropy_loss = -tf.math.reduce_mean(target_actions_logp_time_major)
+
+        # The summed weighted loss.
+        total_loss = (
+            mean_pi_loss
+            + (mean_vf_loss * self.config["vf_loss_coeff"])
+            + (mean_entropy_loss * self.entropy_coeff)
+            + (mean_kl_loss * self.kl_coeff)
+        )
+
+        self.stats = {
+            "total_loss": total_loss,
+            "policy_loss": mean_pi_loss,
+            "vf_loss": mean_vf_loss,
+            "values": values_time_major,
+            "entropy_loss": mean_entropy_loss,
+            "vtrace_adjusted_target_values": vtrace_adjusted_target_values,
+            "mean_kl": mean_kl_loss,
+        }
+        return total_loss
+
+    @override(EagerTFPolicyV2)
+    def stats_fn(self, train_batch: SampleBatch) -> Dict[str, TensorType]:
+        return {
+            "cur_lr": tf.cast(self.cur_lr, tf.float64),
+            "policy_loss": self.stats["policy_loss"],
+            "entropy": self.stats["entropy_loss"],
+            "entropy_coeff": tf.cast(self.entropy_coeff, tf.float64),
+            "var_gnorm": tf.linalg.global_norm(self.model.trainable_variables),
+            "vf_loss": self.stats["vf_loss"],
+            "vf_explained_var": explained_variance(
+                tf.reshape(self.stats["vtrace_adjusted_target_values"], [-1]),
+                tf.reshape(self.stats["values"], [-1]),
+            ),
+            "mean_kl": self.stats["mean_kl"],
+        }
+
+    @override(EagerTFPolicyV2)
+    def make_rl_module(self) -> "RLModule":
+        module = super().make_rl_module()
+
+        # initialize the old_policy
+        # we call this target model so that we can use the target network mixin
+        # however this is not the same as the target policy when we are referring to
+        # target and behavior policies.
+        # additionally, out appo rl module contains its own target network, however to
+        # maintain compatibilty with the target network mixin, we instead are creating
+        # a duplicate rl module here, and won't be utilizing that network.
+        self.target_model = super().make_rl_module()
```

Comment:
wow wow wow. This is a big no no. Only one super call. Just deep copy module into self.target_model man!!